engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/StaticAsOfJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,390 @@
+            ColumnSource<?>[] keySourcesForErrorMessages, int tableSize, double maximumLoadFactor) {
+        // region super
+        super(keySourcesForErrorMessages);
+        // endregion super
+
+        this.tableSize = tableSize;
+        Require.leq(tableSize, "tableSize", MAX_TABLE_SIZE);
+        Require.gtZero(tableSize, "tableSize");
+        Require.eq(Integer.bitCount(tableSize), "Integer.bitCount(tableSize)", 1);
+        Require.inRange(maximumLoadFactor, 0.0, 0.95, "maximumLoadFactor");
+
+        // region constructor
+        mainKeySources = new WritableColumnSource[tableKeySources.length];
+        chunkTypes = new ChunkType[tableKeySources.length];
+        leftRowSetSource = new ImmutableObjectArraySource<>(RowSetBuilderSequential.class, null);
+        rightRowSetSource = new ImmutableObjectArraySource<>(Object.class, null);
+        // endregion constructor
+
+        for (int ii = 0; ii < tableKeySources.length; ++ii) {
+            chunkTypes[ii] = keySourcesForErrorMessages[ii].getChunkType();
+            mainKeySources[ii] = InMemoryColumnSource.getImmutableMemoryColumnSource(tableSize,
+                    tableKeySources[ii].getType(), tableKeySources[ii].getComponentType());
+        }
+
+        this.maximumLoadFactor = maximumLoadFactor;
+
+        ensureCapacity(tableSize);
+    }
+
+    private void ensureCapacity(int tableSize) {
+        for (WritableColumnSource<?> mainKeySource : mainKeySources) {
+            mainKeySource.ensureCapacity(tableSize);
+        }
+        // region ensureCapacity
+        leftRowSetSource.ensureCapacity(tableSize);
+        rightRowSetSource.ensureCapacity(tableSize);
+        // endregion ensureCapacity
+    }
+
+    BuildContext makeBuildContext(ColumnSource<?>[] buildSources, long maxSize) {
+        return new BuildContext(buildSources, (int) Math.min(CHUNK_SIZE, maxSize));
+    }
+
+    ProbeContext makeProbeContext(ColumnSource<?>[] buildSources, long maxSize) {
+        return new ProbeContext(buildSources, (int) Math.min(CHUNK_SIZE, maxSize));
+    }
+
+    static boolean addIndex(ImmutableObjectArraySource<RowSetBuilderSequential> source, long location, long keyToAdd) {
+        boolean addedSlot = false;
+        RowSetBuilderSequential builder = source.getUnsafe(location);
+        if (builder == null) {
+            source.set(location, builder = RowSetFactory.builderSequential());
+            addedSlot = true;
+        }
+        builder.appendKey(keyToAdd);
+        return addedSlot;
+    }
+
+    /**
+     * Returns true if this is the first left row key added to this slot.
+     */
+    protected boolean addLeftIndex(long tableLocation, long keyToAdd) {
+        return addIndex(leftRowSetSource, tableLocation, keyToAdd);
+    }
+
+    protected void addRightIndex(long tableLocation, long keyToAdd) {
+        // noinspection unchecked
+        addIndex((ImmutableObjectArraySource) rightRowSetSource, tableLocation, keyToAdd);
```

Comment:
why the cast?

Reply:
It seems to be needed, chunked version has this:
`    private void addRightIndex(long tableLocation, long keyToAdd) {
        //noinspection unchecked
        addIndex((ObjectArraySource) rightRowSetSource, tableLocation, keyToAdd);
    }`